gemini/sample-apps/accelerating_product_innovation/app/pages_utils/utils_resources_store_embeddings.py

```diff
@@ -0,0 +1,477 @@
+"""
+This module provides functions for processing uploaded files, generating text
+embeddings,
+and managing data within the Google Cloud Storage (GCS) bucket.
+
+This module:
+    * Parses different file formats (CSV, text, Word, PDF), extracts
+    text, splits it into chunks, and creates data packets.
+    * Leverages `embedding_model_with_backoff` to embed text chunks.
+    * Uploads processed data packets to a GCS bucket.
+    * Stores embeddings alongside their associated metadata.
+"""
+
+# pylint: disable=E0401
+
+import asyncio
+import json
+import logging
+import os
+from typing import Any
+
+from PyPDF2 import PdfReader
+import aiohttp
+from app.pages_utils.embedding_model import embedding_model_with_backoff
+import docx
+from dotenv import load_dotenv
+from google.cloud import storage
+import numpy as np
+import pandas as pd
+import streamlit as st
+
+load_dotenv()
+
+
+logging.basicConfig(format="%(levelname)s:%(message)s", level=logging.DEBUG)
+
+
+PROJECT_ID = os.getenv("PROJECT_ID")
+LOCATION = os.getenv("LOCATION")
+
+
+def get_chunks_iter(text: str, maxlength: int) -> list[str]:
+    """Gets the chunks of text from a string.
+
+    This function gets the chunks of text from a string.
+    It splits the string into chunks of the specified maximum length and
+    returns a list of the chunks.
+
+    Args:
+        text (str): The string to get the chunks of.
+        maxlength (int): The maximum length of the chunks.
+
+    Returns:
+        list[str]: A list of the chunks of text.
+    """
+    start = 0
+    end = 0
+    final_chunk = []
+    while start + maxlength < len(text) and end != -1:
+        end = text.rfind(" ", start, start + maxlength + 1)
+        final_chunk.append(text[start:end])
+        start = end + 1
+    final_chunk.append(text[start:])
+    return final_chunk
+
+
+def create_data_packet(
+    file_name: str,
+    page_number: int,
+    chunk_number: int,
+    file_content: str,
+) -> dict:
+    """Creates a data packet.
+
+    This function creates a data packet.
+    It takes the file name, page number, chunk number, and file content as
+    input and returns a dictionary with the data packet.
+
+    Args:
+        file_name (str): The name of the file.
+        page_number (int): The page number of the chunk.
+        chunk_number (int): The chunk number of the chunk.
+        file_content (str): The content of the chunk.
+
+    Returns:
+        dict: A dictionary with the data packet.
+    """
+    # Creating a simple dictionary to store all information
+    # (content and metadata) extracted from the document
+
+    data_packet = {}
+    data_packet["file_name"] = file_name
+    data_packet["page_number"] = str(page_number)
+    data_packet["chunk_number"] = str(chunk_number)
+    data_packet["content"] = file_content
+    return data_packet
+
+
+async def add_type_col(pdf_data: pd.DataFrame) -> pd.DataFrame:
```

Comment:
Does this need to be a dedicated function?

Reply:
Keeping this as a dedicated function for parallel calling